### PR TITLE
BlockEditor: Add BlockCanvas component = Iframe + BlockList + WritingFlow

### PR DESCRIPTION
**Pull request by youknowriad**
Related #53874

## What and why?

Gutenberg can be used as a platform/framework to build block editors. Mainly thanks to the @wordpress/block-editor package. That said, the experience today is not as straightforward as it can be. There can be a lot of small gotchas and hacks you need to do in order to achieve the desired result. One of these small things is the need to manually render `WritingFlow`, `BlockList` and potentially `Iframe` component. this PR proposes the addition of a new component called `BlockCanvas` whose purpose is to bundle all of these into a single component that represents the "canvas" of the editor. 

Principles:

 - Using `<BlockCanvas />` should be enough to render the block list and have all the keyboard behaviors setup (writing flow, is typing flag...)
 - By default it should render the block list within an iframe and should work without any tweaks to styles...
 - Should provide a way to inject styles into the canvas (shortcut to `EditorStyles` component) using a `styles` prop.

Todo and open questions:

 - [x] Ideally the `iframeProps` and `contentRef` props should private to start with. We need to check whether these behaviors can be bundled or replaced with more meaningful props.
 - [ ] I've noticed in the playground that some styles are not loading properly within the iframe by default (for instance the block placeholders and the block appender appear unstyled). Also, the in-between inserter doesn't seem to be positioned properly. (Maybe because of the double iframe)
 - [x] Right now I copied the default styles from the post editor but I also noticed that the iframe component offer props like "expand", "scale", "frameSize". All of these are used for the same purpose: style the iframe width/height and paddings properly. I think ideally the BlockCanvas size and position should be controlled using meaningful props rather than random styles built-in or a random "style" prop. So we should have a good default (maybe the block canvas auto expand its height as you type and the width should be 100% by default) while allowing customization using semantic props (like expand...).
 - [x] Document the component properly.

## Testing Instructions

1- E2E tests should pass.

### CODE REVIEW

File: packages/block-editor/README.md [hunk 1]
```diff
@@ -118,6 +116,10 @@ _Returns_
 
 -   `WPElement`: Block Breadcrumb.
 
+### BlockCanvas
+
+Undocumented declaration.
```
Comment (jorgefilipecosta) [inline packages/block-editor/README.md]: Should we include some documentation for this new component?

File: packages/block-editor/src/components/block-canvas/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import { useMouseMoveTypingReset } from '../observe-typing';
+
+function BlockCanvas( {
+	shouldIframe = true,
+	children = <BlockList />,
+	styles,
+	contentRef,
+	iframeProps,
+} ) {
+	const resetTypingRef = useMouseMoveTypingReset();
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-canvas/index.js:18]: I don't know if the call useMouseMoveTypingReset is expensive in terms of performance. But suppose it has some impact given that it is not needed when shouldIframe is false. In that case, we may consider having an inner internal component wrapping the Iframe and making the useMouseMoveTypingReset call, or even call useMouseMoveTypingReset inside the iframe if it is always needed.

**Review by jorgefilipecosta — APPROVED** [review]
The change looks good I think we can merge it right after resolving the conflict.

**Comment by alexstine** [comment]
It sounds like a good idea. Even knowing how some of these components work, I have a hard time tracing through what does what.

**Comment by youknowriad** [comment]
What should be the default "height" of the new BlockCanvas component?

 - Make it auto-expand by default (like a regular div). This had my preference originally but because of how complex it is for iframes, I decided that maybe we should avoid it for now.
 - 100%: take the height of its container. In WordPress, this is what we use almost everywhere, downside is that third-party users must define a height for the container div otherwise the block canvas will have "height: 0" by default (hidden)
 - A random number like 300px: Feels like this is the most common in editors in the wild (think tinymce...), it's good for use-cases like isolated block editor but rarely used in WordPress.

After several considerations, I'm going with the last option as it's the nicest in terms of combination between API and behavior. By default it works but it also allows you to support the "fit container" use case by passing 100% to the prop. We may support the first use-case at some point by passing "auto" but because of the complexity, I think it's not worth addressing for now.

File: packages/block-editor/src/components/block-canvas/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-canvas/index.js:18]: This hook does nothing really unless you pass it to an element, so it's fine.

**Comment by youknowriad** [comment]
I marked most of the props of the `BlockCanvas` component as private because I think ideally, all usage should use the iframe, and that we're not sure yet about the viability of the contentRef and iframeProps prop. `children`, `height` and `styles` are clearly good props for me, so these are the only ones I'm leaving as public for now.

**Comment by youknowriad** [comment]
I think we're now ready to merge the initial version of this component (intentionally limited in terms of customization options and props).

>  I've noticed in the playground that some styles are not loading properly within the iframe by default (for instance the block placeholders and the block appender appear unstyled). Also, the in-between inserter doesn't seem to be positioned properly. (Maybe because of the double iframe)

@ellatrix you might want to be aware of this, I think we can probably debug that separately though.

**Pull request merged**